File: cpp/lib/rsl/TokenList.cpp
#include "rsl/TokenList.hpp"

#include <stdexcept>

namespace rsl
{
    TokenList& TokenList::ExpectFront(const TokenType tokenType)
    {
        if (const auto& a = Front(); a.type != tokenType)
        {
            throw std::runtime_error("Unexpected token");
        }

        return *this;
    }

    TokenList& TokenList::ExpectBack(const TokenType tokenType)
    {
        if (const auto& a = Back(); a.type != tokenType)
        {
            throw std::runtime_error("Unexpected token");
        }

        return *this;
    }

    TokenList& TokenList::ExpectFront(const std::vector<TokenType>& tokenTypes)
    {
        if (const auto& a = Front(); std::find(tokenTypes.begin(), tokenTypes.end(), a.type) != tokenTypes.end())
        {
            throw std::runtime_error("Unexpected token");
        }

        return *this;
    }

    TokenList& TokenList::ExpectBack(const std::vector<TokenType>& tokenTypes)
    {
        if (const auto& a = Back(); std::find(tokenTypes.begin(), tokenTypes.end(), a.type) != tokenTypes.end())
        {
            throw std::runtime_error("Unexpected Token");
        }

        return *this;
    }

    Token TokenList::RemoveFront()
    {
        // Call for checks
        auto a = Front();
        _tokens.pop_front();
        return a;
    }

    Token TokenList::RemoveBack()
    {
        // Call for checks
        auto a = Back();
        _tokens.pop_back();
        return a;
    }

    Token& TokenList::Front()
    {
        if (Empty())
        {
            throw std::runtime_error("Expected Input");
        }
        return _tokens.front();
    }

    Token& TokenList::Back()
    {
        if (Empty())
        {
            throw std::runtime_error("Expected Input");
        }
        return _tokens.back();
    }

    TokenList& TokenList::InsertFront(const Token& token)
    {
        _tokens.push_front(token);
        return *this;
    }

    TokenList& TokenList::InsertBack(const Token& token)
    {
        _tokens.push_back(token);
        return *this;
    }

    bool TokenList::Empty() const
    {
        return _tokens.empty();
    }

    bool TokenList::NotEmpty() const
    {
        return !Empty();
    }
}

File: csharp/rsl/TokenList.cs
namespace rsl;

public sealed class TokenList
{
    private readonly LinkedList<Token> _list = new();
    private Token? _lastToken = null;

    public Exception CreateException(string message, Token token)
    {
        return new ExceptionWithDebug(token.DebugInfo,message);
    }

    public void ThrowExpectedInput()
    {
        if(_lastToken != null){
            throw new ExceptionWithDebug(_lastToken.DebugInfo,"Expected Input");
        }
        throw new Exception("Expected Input");
    }

    public Token RemoveFront()
    {
        if (Empty()) ThrowExpectedInput();

        var a = Front();
        _list.RemoveFirst();
        if(Empty()){
            _lastToken = a;
        }
        /*if (NotEmpty())
        {

        }*/
        return a;
    }

    public Token RemoveBack()
    {
        if (Empty()) ThrowExpectedInput();

        var a = Back();
        _list.RemoveLast();
        _lastToken = a;
        return a;
    }

    // public TokenList ExpectFront(TokenType type)
    // {
    //     if (Empty())
    //     {
    //         ThrowExpectedInput();
    //     }
    //
    //     var a = Front();
    //     
    //     if(a.Type != type) throw CreateException("Unexpected input",a);
    //     return this;
    // }

    public TokenList ExpectFront(params TokenType[] type)
    {
        if (Empty()) ThrowExpectedInput();

        var a = Front();

        if (!type.Contains(a.Type)) throw CreateException("Expected " + "[" + type.Aggregate("",(c,d) => $" {d.ToString()} ") + "]" + $" but got {a.Type.ToString()}", a);
        return this;
    }

    // public TokenList ExpectBack(TokenType type)
    // {
    //     if (Empty())
    //     {
    //         ThrowExpectedInput();
    //     }
    //
    //     var a = Back();
    //     
    //     if(a.Type != type) throw CreateException("Unexpected input",a);
    //     return this;
    // }
    public TokenList ExpectBack(params TokenType[] type)
    {
        if (Empty()) ThrowExpectedInput();

        var a = Back();

        if (!type.Contains(a.Type)) throw CreateException("Expected " + "[" + type.Aggregate("",(c,d) => $" {d.ToString()} ") + "]" + $" but got {a.Type.ToString()}", a);
        return this;
    }

    public Token Front()
    {
        if (Empty()) ThrowExpectedInput();

        return _list.First();
    }

    public Token Back()
    {
        if (Empty()) ThrowExpectedInput();

        return _list.Last();
    }

    public int Size()
    {
        return _list.Count;
    }

    public TokenList InsertFront(TokenList other)
    {
        while (other.NotEmpty())
        {
            _list.AddFirst(other.RemoveBack());
        }

        return this;
    }

    public TokenList InsertFront(Token token)
    {
        _list.AddFirst(token);
        return this;
    }

    public TokenList InsertBack(TokenList other)
    {
        while (other.NotEmpty())
        {
            _list.AddLast(other.RemoveFront());
        }

        return this;
    }

    public TokenList InsertBack(Token other)
    {
        _list.AddLast(other);
        return this;
    }

    public TokenList Clear()
    {
        _list.Clear();
        return this;
    }

    public bool Empty()
    {
        return Size() == 0;
    }

    public bool NotEmpty()
    {
        return Size() > 0;
    }
}